src/lib.rs

```diff
@@ -89,3 +89,5 @@ pub mod frame;
 pub use block::{compress, compress_into, compress_prepend_size};
 
 pub use block::{decompress, decompress_into, decompress_size_prepended};
+
+pub(crate) mod sink;
```

Comment:
I think this module violates the no-unsafe guarantee for safe-encode / safe/decode

Reply:
Strictly speaking, yes. In hides the unsafety behind a safe api. Much like a Vec. This is a trade-off so a lot of code may be reused. The upside is that the unsafe safety of the sink module is easily verifiable.

Reply:
MaybeUninit may also disable possible optimizations in safe code. 
When Sink is generic over the MaybeUninit or a slice it would be safe.
But not sure how big the api is

Reply:
Maybe, but it seems unlikely considering that that the contents of the `Sink` never participate in any branches/comparisons.

Reply:
What do you think about a trait used in `Sink`. The api is not large, although the discriminator seems to be larger than theoretically necessary (generic over the slice type not the whole datastructure)

```
/// MaybeUninitOrMaybeNot is an abstraction over sometimes uninit data.
/// It is either &[u8] or &[MaybeUninit<u8>]
///
trait MaybeUninitOrMaybeNot {
    fn set(&mut self, byte: u8, pos: usize);
    fn fill(&mut self, range: Range<usize>, byte: u8);
    fn copy_from_slice(&mut self, range: Range<usize>, slice: &[u8]);
    fn copy_within_(&mut self, range: Range<usize>, dest: usize);
}

impl MaybeUninitOrMaybeNot for &mut [u8] {
    fn set(&mut self, byte: u8, pos: usize) {
        self[pos] = byte;
    }

    fn fill(&mut self, range: Range<usize>, byte: u8) {
        self[range].fill(byte);
    }

    fn copy_from_slice(&mut self, range: Range<usize>, slice: &[u8]) {
        self[range].copy_from_slice(slice);
    }

    fn copy_within_(&mut self, range: Range<usize>, dest: usize) {
        self.copy_within(range, dest);
    }
}

impl MaybeUninitOrMaybeNot for &mut [MaybeUninit<u8>] {
    fn set(&mut self, byte: u8, pos: usize) {
        self[pos] = MaybeUninit::new(byte);
    }

    fn fill(&mut self, range: Range<usize>, byte: u8) {
        self[range].fill(MaybeUninit::new(byte));
    }

    fn copy_from_slice(&mut self, range: Range<usize>, data: &[u8]) {
        self[range].copy_from_slice(slice_as_uninit_ref(data));
    }

    fn copy_within_(&mut self, range: Range<usize>, dest: usize) {
        self.copy_within(range, dest);
    }
}
```

Reply:
That might work it just needs the "wild" variants. Is the goal to have a non unsafe version?

Reply:
At some point I had a trait version that benchmarked the same as the version I have in this PR. You can see it here https://github.com/PSeitz/lz4_flex/compare/master...arthurprs:sinktrait?expand=1#diff-62de54f049272daf34880594f760f7cb4ee9efdf5acd2b4683660a69a16ecaf3R13-R24

It illustrates that such an API is possible, but banishing unsafe is a road of diminish returns :smile:

Reply:
> ~That might work, it just needs the "wild" variants.~ Edit: nevermind this, I understand now.
> 
> Is the goal to have a non unsafe version?

ah sorry for my late reply, I'll have more time now.

Yep, I just think it would be nice to keep the no-unsafe promise, if the feature flag is not enabled. The whole `MaybeUninit` stuff shows that `unsafe` can have unexpected side effects. If someone has for whatever reason the need to have bullet proof code, I think it would be nice to provide that.

You mentioned diminishing returns, do you mean performance wise? The mentioned commit also has `unsafe` blocks.

Reply:
> The mentioned commit also has unsafe blocks.

Yes, it was just an illustration that using a trait might work. I think with some work the aforementioned branch can provide a fully safe implementation of `impl Sink for &[u8]`. Then the safe encoder/decoder can pre-fill a Vec and build a `Sink<&[u8]>` with the initialized bytes.

I'll try to find the time to update that branch to provide that.

> You mentioned diminishing returns

I mean more in terms of code maintenance, I noticed things are getting increasingly more complicated in order to provide the _100%_ safe variants. But I agree it's something people want in the rust ecosystem.